### ISSUE TITLE
Adam (and other adaptive algorithms) use epsilon the wrong way

### ISSUE THREAD

**Issue by pandrey-fr** [keras-team/tf-keras#315]
**System information**.
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.11
- Python version: 3.10
- Problem comes from the source code algorithm

**Describe the problem**.

The current implementation of Adam (and, in fact, of other adaptive algorithms such as Adagrad or RMSprop) is making a use of the `epsilon` hyper-parameter that is inconsistent with the literature ([Adam paper](https://arxiv.org/pdf/1412.6980.pdf), Algorithm 1), with other libraries (such as Torch and MXNET), and with the rationale behind it.

In the paper, the update rule for Adam is stated to be $\theta_t = \theta_{t-1} - \eta_t * \hat{m_t} / (\sqrt{\hat{v_t}} + \epsilon)$, but the current Keras implementation instead applies $\theta_t = \theta_{t-1} - \eta_t * \hat{m_t} / (\sqrt{\hat{v_t} + \epsilon})$.

**Edit**: my statement about Adam is wrong (the problem is not with the place of epsilon the computation but with $\sqrt{\hat{v_t}}$ - see the discussion in comments below), however I stand by it for the more general addition below:

More generally, for algorithms such as Adagrad or RMSprop, when the update rule is mathematically stated to be $\theta_t = \theta_{t-1} - \eta_t * \textnormal{num} / \sqrt{\textnormal{den}}$, it is common practice to add an epsilon term to the denominator in order to prevent division by zero and/or numerical under or overflow. The practice in most frameworks, which in my humble opinion is coherent with the former rationale, is then to use $\theta_t = \theta_{t-1} - \eta_t * \textnormal{num} / (\sqrt{\textnormal{den}} + \epsilon)$, hence considering that the square root is anecdotal, and that what matters is that the final denominator is greater or equal than epsilon. However, in keras, there is a consistent use of $\theta_t = \theta_{t-1} - \eta_t * \textnormal{num} / (\sqrt{\textnormal{den} + \epsilon})$.

**Describe the current behavior**.

As stated above, formulas make use of $\theta_t = \theta_{t-1} - \eta_t * \textnormal{num} / (\sqrt{\textnormal{den} + \epsilon})$.

**Describe the expected behavior**.

As stated above, formulas should make use of $\theta_t = \theta_{t-1} - \eta_t * \textnormal{num} / (\sqrt{\textnormal{den}} + \epsilon)$, out of consistency with other reference implementations, the literature (in the case of Adam), and the rationale behind the use of epsilon.

Please note that one cannot just change the value of the hyper-parameter to achieve a behaviour similar to that of other frameworks, since the adjusted epsilon would be dependent on the value of the base denominator, which is typically a function of the input gradients and/or a stateful variable (momentum, velocity...).

At the very least, if changing the formula is not deemed an adequate solution, I think that it could be worth documenting this point of discrepancy between keras and other frameworks, as it can have non-trivial effects on the reproducibility of experiments depending on the chosen backend framework.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no): yes
- Briefly describe your candidate solution(if contributing): modify the backend of adaptive optimizers to move epsilon out of the square root of the denominator term it is added to

**Standalone code to reproduce the issue**.

There is no code to run, but rather code to review. For Adam:
- [keras source code](https://github.com/keras-team/keras/blob/v2.11.0/keras/optimizers/optimizer_experimental/adam.py#L157-L200)
- [torch source code](https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L49-L85)
- [mxnet source code](https://github.com/apache/mxnet/blob/master/python/mxnet/optimizer/adam.py#L106-L147)
- [reference paper](https://arxiv.org/abs/1412.6980)

**Comment by carlosgmartin** [comment]
[optax](https://github.com/google-deepmind/optax) disambiguates this by using separate `epsilon` and `epsilon_root` parameters in optimizers.

A possible solution here is to add parameters `epsilon_inner` and `epsilon_outer`, denoting epsilons for inside and outside the square root, respectively, and to deprecate plain `epsilon` as ambiguous.
